mesonbuild/dependencies/base.py

```diff
@@ -317,8 +317,14 @@ def get_variable(self, *, cmake: T.Optional[str] = None, pkgconfig: T.Optional[s
         raise DependencyException(f'Could not get an internal variable and no default provided for {self!r}')
 
     def generate_link_whole_dependency(self) -> Dependency:
+        from ..build import SharedLibrary
         new_dep = copy.deepcopy(self)
-        new_dep.whole_libraries += new_dep.libraries
+        # XXX: we need to check that CustomTargets are not SharedLibs...
+        if any(isinstance(x, SharedLibrary) for x in new_dep.libraries):
+            raise MesonException(f'Cannot convert a dependency to link_whole with shared libraries')
```

Comment:
I guess technically this could be an InvalidArguments, but it may not particularly matter?